codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java

```diff
@@ -280,7 +280,11 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
                 }
 
                 int size = out.size();
-                decodeWasNull = !out.insertSinceRecycled();
+                if (decodeWasNull) {
+                    // Only update decodeWasNull if we did not produce data in a previous channelRead(...) call.
+                    decodeWasNull = !out.insertSinceRecycled();
```

Comment:
consider avoiding a conditional/branch statement here and just do the assignment.

```java
decodeWasNull = decodeWasNull && !out.insertSinceRecycled();
```